src/compiler/sys.ts

```diff
@@ -297,6 +304,131 @@ namespace ts {
                 };
             }
 
+            function createWatchedFileSet() {
+                const dirWatchers = createFileMap<DirWatcher>();
+                const recursiveDirWatchers = createFileMap<DirWatcher>();
+                const fileWatcherCallbacks = createFileMap<FileWatcherCallback>();
+                const dirWatcherCallbacks = createFileMap<DirWatcherCallback>();
+
+                const currentDirectory = process.cwd();
+                return { addFile, removeFile, addDir };
+
+                function addDir(dirName: string, callback: DirWatcherCallback, recursive?: boolean) {
+                    const dirPath = toPath(dirName, currentDirectory, getCanonicalPath);
+                    dirWatcherCallbacks.set(dirPath, callback);
+                    const { watcher, isRecursive } = addDirWatcher(dirPath, recursive);
+                    return {
+                        close: () => reduceDirWatcherRefCount(watcher, dirPath, isRecursive)
+                    };
+                }
+
+                function reduceDirWatcherRefCount(watcher: DirWatcher, dirPath: Path, isRecursive: boolean) {
+                    watcher.referenceCount -= 1;
+                    if (watcher.referenceCount <= 0) {
+                        watcher.close();
+                        if (isRecursive) {
+                            recursiveDirWatchers.remove(dirPath);
+                        }
+                        else {
+                            dirWatchers.remove(dirPath);
+                        }
+                    }
+                }
+
+                function addDirWatcher(dirPath: Path, recursive?: boolean): { watcher: DirWatcher, isRecursive: boolean } {
+                    let watchers: FileMap<DirWatcher>;
+                    const options: { persistent: boolean, recursive?: boolean } = { persistent: true };
+
+                    // Node 4.0 `fs.watch` function supports the "recursive" option on both OSX and Windows
+                    // (ref: https://github.com/nodejs/node/pull/2649 and https://github.com/Microsoft/TypeScript/issues/4643)
+                    if (isNode4OrLater() && recursive === true) {
+                        if (recursiveDirWatchers.contains(dirPath)) {
+                            const watcher = recursiveDirWatchers.get(dirPath);
+                            watcher.referenceCount += 1;
+                            return { watcher, isRecursive: true };
+                        }
+                        watchers = recursiveDirWatchers;

```

Comment:
not sure that I got this right: here we'll alias `recursiveDirWatchers` as `watchers` and then at the line 365 we'll put fresh watcher in `watchers` so it will end up in `recursiveDirWatchers` however on the next line we'll return `{ watcher, isRecursive: false}`. Since we are using value of  `isRecursive` to pick the map to delete the element it seems we'll use wrong set.